### ISSUE TITLE
Foundation 6: HTML/Sass feedback

### ISSUE THREAD

**Issue by gakimball** [zurb/foundation#6196]
In this thread, we'd like you hear your feedback on how HTML, Sass, and CSS is written in Foundation for Sites.

## Accessibility

About half of the Foundation framework is currently keyboard/AT accessible. The goal for version 6 is 100%, so we're writing every component from the ground up with accessibility in mind. Of course, the mixin-based strucutre of our components means the user will have plenty of flexibility when it comes to HTML, but in our documentation we'll prescribe a recommended way of writing accessible markup.

## Class style

Foundation has never enforced a rigid style for writing and structuring CSS classes. In fact, our mixins allow you to use our components in BEM, SMACSS, et al. if you want. That being said, we have heard a few things repeatedly about how our CSS could be structured.

### Tag vs. class selectors

Foundation has an even mix of tag and class selectors. For some components, child elements are selected with tags (e.g., button groups, which use `.button-group` for the container and `.button-group > li` for children), while others are selected with classes (e.g., accordions, which use `.accordion` and `.accordion-navigation`). We've had some requests to shift to a more heavily class-based approach to writing styles.

There are a few benefits to a class-based approach, including:
 - Less specificity, due to a reduced need for nesting selectors. A common criticism of large frameworks is that it's difficult to override the default styles, and this is a step in that direction, if the user doesn't have to look up (or guess) the specificity of an element.
 - Less potential for collision between components. Users often run into trouble when trying to put one component inside of another. If the parent component includes descendant tag selectors, those will cascade down to the inner component, disrupting its styling.

The knocks against it would be:
 - More verbosity. A heavier (or entirely) class-based approach means longer class names in general. If a component is a list of items, writing `<li class="component-item">` is a much bigger hassle than just `<li>`.
 - Not leveraging HTML semantics. "Separate style from structure" is a common mantra in CSS, which is why people avoid tag selectors, but tags exist for a reason. This is more important for Foundation 6, because it's going to be accessible from the ground up, and accessible markup can't just be a sea of `<div>`s.

Heydon Pickering wrote [a great article](http://alistapart.com/article/axiomatic-css-and-lobotomized-owls) on A List Apart that briefly touches on this. Read the comments as well to see more thoughts of his.

## The Settings File

While Foundation for Apps was in development, we had a [discussion](https://github.com/zurb/foundation-apps/issues/23) on ways to make the settings file more useful. Right now Foundation (and most other customizable UI frameworks) dumps all of its settings into a single file as individual variables. There's no concept of namespacing or structure, other than the structure we enforce by how we write the code. It's served us well so far, and people don't really complain about it, but that doesn't mean there isn't a better way.

For example, to configure the Sass grid framework Susy, you create a `$susy` map which will override the default values.

```scss
$susy: (
  columns: 12,
  gutters: 1/4,
  math: fluid,
  output: float,
  gutter-position: inside,
);
```

The problem with Foundation is that it has a good two dozen components, which makes your map a lot bigger:

```scss
$foundation: (
  global: (
    rem-base: 16px,
    primary-color: #2ca9cd,
  ),
  grid: (
    column-count: 12,
    column-gutter: rem-calc(30),
  ),
);
```

We could also streamline this by using individual variables for each component, and then merging them in the background.

```
$foundation-grid: (
  column-count: 12,
  column-gutter: rem-calc(30),
);
```

There's also an issue of access. Maps have been around for a while now in Sass, but I can tell you from experience there are plenty of more casual users of Sass who wouldn't know what they are, much less what an "array" is, since they aren't programmers. Maps feel like more of a Sass developer feature than a Sass user feature. Here are some ways variable access might work:

```scss
.component {
  // The current way
  color: $primary-color;

  // With a pure map function
  color: map-get($foundation-global, primary-color);

  // With an abstraction function
  color: foundation(global, primary-color);
}
```

The question is if that is any more or less difficult to access than a typical Sass variable. To be sure, the vast majority of Sass variables are never used by developers. It's mostly global styles like `$primary-color` that gets used a lot, so we don't want to confuse people too much with new syntax.

**Here's what we'd like to hear:**
  - Should we go all-in on verbose class names and shallow nesting, or keep things how they are with some slight improvements?
  - Should we enforce a specific naming convention, like BEM or SMACSS? Traditionally, Foundation's class names haven't adhered to any strict guidelines.
  - How can we improve the settings file?

**[View our other discussion topics for Foundation 6 here.](https://github.com/zurb/foundation/labels/Foundation%206)**

**Comment by mjatharvest** [comment]
I wouldn't want to see the settings file push the limits to the point it wouldn't be compatible with libsass, it's great not to have a ruby dependency

**Comment by gakimball** [comment]
@mjatharvest Libsass 3.1 has full map compatibility, so we're good there. There are little things still missing that would be useful (like `@at-root`), but most of the really new Sass stuff (selector functions) we can live without.

**Comment by izakfilmalter** [comment]
Would love to see BEM. You would gain performance and also reusability. If you want to do a bunch of map in map stuff in sass, you will run into the list issue. A sass map can not have another map as the value. It can have a list which inturn can have a map in. Had to deal with this in https://github.com/ModernInc/modern-style. I would aim for the removal of nesting completely. Ask yourself what you gain with nesting vs a unique class name. If your gonna follow BEM follow it all the way. Don't half follow it with only naming.

Finally, foundation 5 is heavier than bootstrap 3. This saddens me because foundation in it's nature should be lighter than bootstrap. Maybe we can work together on this but using something like the Modern Map API that I built to allow unique elements and blocks to share styles could help shrink the bloat present.

**Comment by gakimball** [comment]
@izakfilmalter To be sure, the main thing you gain from less classes is *convenience*, but I don't know if that makes it a good thing. Given a menu bar component like this:

```html
<ul class="menu-bar">
  <li><!-- ... --></li>
  <li><!-- ... --></li>
</ul>
```

Just writing a bunch of `<li>`s, a tag which already means "list item", is easier than writing `<li class="menu-bar__item">`. A child selector ensures that the styles don't leak into any nested lists inside the component. It may restrict you to using only one kind of tag for this element, but the vast majority of tags are just boxes from a technical perspective, so as long as we're using the tags for their semantic purpose, you can style it the same as you would a `<div>`.

But like I said, it's mostly a convenience thing, and it has created problems for us in the past when our selectors weren't airtight.

**Comment by izakfilmalter** [comment]
@gakimball I mean most people should be using emmet by now. We could even pitch that foundation works well with it. `ul.menu-bar>li.menu-bar__item*6` and you're done.

**Comment by gakimball** [comment]
@izakfilmalter True that, and I'm also ignoring the fact that with dynamically-generated content it's less of a problem. It's more of an issue when coding static websites.

**Comment by izakfilmalter** [comment]
@gakimball but emmet. Like in all reality we should be pushing both a framework that saves time and then awesome tools that can be combined with it to save more time.

**Comment by ksherman** [comment]
I'm 100% in favor maps for settings, grabbing a new version and trying to hunt down my customizations is little frustrating, even if I have to learn a little bit.

**Comment by joelkinzel** [comment]
The more I've used BEM the more I've seen it's benefits. Even if it isn't straight BEM, some sort of convention should be used.

I'm also a huge proponent of moving to more verbose classes. Yes, it means you have to potentially add more classes to things, BUT it also allows flexibility in the way they are built (i.e. styles are not tied to specific HTML elements). One thing that really irked me, and where most of my customization needs have come from is from the CSS and JS being too specific and breaking when I change the HTML element (i.e. I don't think DL is an appropriate element semantically for accordions). 

Here is another thought on how to organize the SASS/CSS for a project from Harry Roberts (if you haven't read his stuff, you really should): http://vimeo.com/114965689

**Comment by navidkashani** [comment]
+1 for class-based approach, it much better on performance.

I love to see foundation use [Sass Guidelines](http://sass-guidelin.es/) or [CSS Guidelines](http://cssguidelin.es/).

Using [scss-lint](https://github.com/causes/scss-lint) and include config file can be very useful for writing Scss like foundation style.

**Comment by johnwbaxter** [comment]
I'd welcome BEM or SMACSS for sure, so +1 me on that!

**Comment by gakimball** [comment]
So internally we've done some messing around with maps as Foundation settings, and there are some interesting things we could do with it. Here's what the syntax might look like:

```scss
// Define component settings
$foundation-button: (
  color: #fff,
  background: (
    default: #3888a2,
    hover: #4ec2e7,
    active: #4ec2e7,
  ),
);

// Add to Foundation
@include foundation-component(button, $foundation-button);

// Access using the foundation() function
.button {
  background: foundation(button background);

  &:hover {
    background: foundation(button background hover);
  }
}
```

There's some really interesting stuff we could do with this, like allowing third-party plugins to hook into the core of the framework, exporting the settings file as JSON (for theme distribution), and more.

There's one huge caveat though, which is that you can't reference previous values in a map while it's being defined.

```scss
// Nope
$map: (
  one: #000,
  two: lighten(map-get($map, one), 10%),
);
```

This makes it way harder to easily theme components. In Foundation a lot of visual styles are inherited or manipulated. For example, you'd want a button's hover state, by default, to be a lightened/darkened version of the default state. But if you're in the middle of defining a `$button` map, you can't reference an earlier value in the map, because it hasn't been created yet. You'd have to go back and merge it in manually.

```scss
$button: (
  background: blue,
);

$button: map-merge($button, lighten(map-get($button, background), 10%));
```

Last night I messed around with creating some ridiculous "shorthand" syntax that could get around this, but it looks really bad. Basically you'd pass this map through a function that adds each property to the map, one by one, so that you can actually backreference things.

```scss
// Call lighten($val, 20%), where $val is this map's 'background' property
$foundation-button: (
  background: blue,
  background-hover: this 'background' lighten 20%,
);
```

It does what it's supposed to do, but the syntax looks ridiculous because you can't use actual function calls, you have to abstract them out instead. It's not really a viable solution at all.

If anyone has any thoughts or ideas on this, we'd love to hear them!

**Comment by coreysyms** [comment]
This request seems like quite a lift but based off Heydon Pickering's A List Apart article and my own experiences. Would breaking out the css into respective "types" files be an option? In addition to utilizing some of the bloat reduction with axiomatic CSS.

Lets say we have:
type.css - Rules for  h1 - h6 p etc.
grid.css - Rules for the grid
input.css - Rules for form objects
ui.css - Rules for UI objects
plugins.css - Rules for all the other fun stuff that Foundation provides

Then include them in foundation.css. It would be a lot easier to understand and help with finding redundant styles or just selectors in general. I find myself pulling in foundation.css then creating my own app.css file to override foundations styles where I need them, this is both by design and also because sifting through foundations CSS file is a task. Foundation.css is 6000 lines. It could also be a type of "plug and play" for CSS, where you remove what you don't need, lets say I want my type to be different, which is almost always the case, I can simply remove the include for type.css and write in my own without overriding foundations styles, slightly speeding up things.

I'm of the mindset that organization and thought goes further than additional technology to compensate for perceived speed. Anyone else find themselves thinking this way?

**Comment by gakimball** [comment]
@coreysyms We tend to recommend using Foundation with Sass if you want the most control over the framework. The plain CSS version is more meant for convenience (getting started quickly without any setup) and portability (using it on a CDN, CodePen, etc.), which is why it's just one big file. (It's also one big file to cut down on HTTP requests.) We at ZURB feel pretty strongly that the use of a preprocessor is essential to a modern web development workflow.

The Sass version gives you that level of control, because every component is separated into individual files that you can easily remove from the finished CSS. Every core visual and structural style is also exposed as a Sass variable that can be changed to whatever you want.

All that to say, we definitely agree with you, to the extent that this is how we organize the Foundation Sass files: modularly, by component.

**Comment by christoffervivian** [comment]
Less SCSS variables. They're mostly useless. Variables should be used for complex things, not simple things like change a link decoration from underline to whatever.

**Comment by fhunt45** [comment]
And please review that all your variables on settings work, because sometimes i have changed few settings (pagination) on settings.scss and saw no changes at all.

**Comment by proimage** [comment]
Why not keep the recommended semantic HTML tag-based approach, but provide backup classes for those cases when either the component is being used in an unanticipated way that is semantically different, or the markup cannot be changed due to closed CMS systems?

For example, using @gakimball's code from above, this could be the recommended, semantic way to use the 'menu-bar' component:

```html
<ul class="menu-bar">
  <li><!-- ... --></li>
  <li><!-- ... --></li>
</ul>
```

However, the selector in the (S)CSS would effectively be:

```scss
.menu-bar
{
    // styling stuff
    > li,
    > .menu-bar__list
    {
        // more styling stuff
    }
}
```

...which  would allow you to achieve the same visual and behavioral result using, say, this code:

```html
<nav class="menu-bar">
  <a href="#" class="menu-bar__list"><!-- ... --></a>
  <a href="#" class="menu-bar__list"><!-- ... --></a>
</nav>
```
In other words, you're providing a tag-based semantic way of coding components wherever possible as the recommended default, _and also_ a class-based fallback for those times when those specific tags simply cannot be used.

**Comment by gakimball** [comment]
@proimage Sometimes the mere existence of tag selectors can create problems, most often when you try to extend existing components. But if we're just smarter about how we write selectors, we can avoid some of that while still using tags.

Here are two examples:

```scss
.menu-bar {
  li { }   // Problematic because any generic <li>s will now be captured
  > li { } // Fixed by using a child selector instead
}
```

```scss
// This component uses a <span> to add an icon
.thing-with-icon {
  span { } // Problematic because it targets spans too generically
  > span { } // Even if you do this, it still means you can't use the <span> tag for any other purpose inside this component
}

**Comment by ahmadmilzam** [comment]
Hello, for F6 i would like to propose 2 points:

1. Naming convention, i would love to chose BEM since it has low specifity because every element inside a module has a class name, so we don't have to nest it unless we need a modifier class
2. Disable every feature and modifier by default, make the base module/component style really plain and modular as possible. i like the OOCSS from nicole sullivan and inuit css from harry robert. I know that F5 has $export-html-classes variable, but it is not enough, i personally still have to remove some rules declaration from the mixin because i don't need that rules.

**Comment by michaelBenin** [comment]
I don't like the fact that you have to use ids in reveal. I'd like to use it with just classnames.

**Comment by gosupernova** [comment]
It is hard to say BEM (class) or tag based as there are benefits for each, we have to remember that this is a framework for people to apply there own styles to so would it make sense that the default is tag based then people can add the BEM structure where if it was the other way around you could not undo it?

I just think soon as you choose one standard you are restricting the framework where people might be looking for a lighter touch rather than such a strict version.

Personally I use BEM but I think this is a user choice not a framework decision

**Comment by rafibomb** [comment]
@christoffervivian We actually discussed this here at ZURB too. We'll work to make the _settings.file less bloated. The challenge as always will be to deny PR's to add more.

**Comment by rafibomb** [comment]
@michaelBenin The new Reveal modal will likely work more like Foundation for Apps http://foundation.zurb.com/apps/docs/#!/modal although with jQuery.

Id's are necessary to attribute what button or link is opening what modal.

**Comment by ksherman** [comment]
I just thought of a pretty big one: spend some time focusing on print output and options. 

Right now, it essentially prints the mobile view which is pretty awful and 16px fonts are way too large. Printing perhaps isn't a huge part of the market, but spending some time on it would be appreciated.

**Comment by gakimball** [comment]
@ksherman We can definitely look into it. Can you think of any examples of websites with really good print styles we could look at?

**Comment by michaelBenin** [comment]
How do we feel about modularizing the settings file?

**Comment by gakimball** [comment]
@michaelBenin Could you elaborate? Do you mean including one settings file for each component?

**Comment by doertedev** [comment]
Another thing: Please don't include files in files. Because path foo and such. Thanks :)

Otherwise: 
```
doertedev@trolovo ~/Code/somefuckinproject/assets $ grunt
Running "compass:dist" (compass) task
    error assets/css/source/project.scss (Line 58 of bower_components/foundation/scss/foundation/_settings.scss: File to import not found or unreadable: foundation/functions.
Load paths:
  Compass::SpriteImporter
  /home/doertedev/Code/somefuckinproject/assets/css/source
  /home/doertedev/.rbenv/versions/2.1.3/lib/ruby/gems/2.1.0/gems/compass-core-1.0.1/stylesheets)
Sass::SyntaxError on line ["58"] of /home/doertedev/Code/somefuckinproject/bower_components/foundation/scss/foundation/_settings.scss: File to import not found or unreadable: foundation/functions.
Load paths:
  Compass::SpriteImporter
  /home/doertedev/Code/somefuckinproject/assets/css/source
  /home/doertedev/.rbenv/versions/2.1.3/lib/ruby/gems/2.1.0/gems/compass-core-1.0.1/stylesheets
Run with --trace to see the full backtrace
Warning: ↑ Use --force to continue.

Aborted due to warnings.
```
Talking about letting the user extend instead of copy things: this is a nono... (or I did oversee smth obvious).

**Comment by proimage** [comment]
One more thing regarding the ´_settings.scss´ - what is the recommended way to modify the settings therein that doesn't get overwritten when upgrading Foundation? I've settled on copying anything I need to modify out of there and into my own custom ´_vars.scss ´. Is this a problem only I encounter because I seem to be the only living web developer on the planet who uses Windows instead of Mac, and thus I don't have access to all that CLI/Bower/NodeJS stuff?

**Comment by doertedev** [comment]
@proimage: What? No. You have access to that bower / cli stuff with windoze.
And: What you do need to do with every update is the same as with every other software you work with: read the Changelogs. Because there you'll find p.e. additional files to include. Or some modules might have changed their name (topbar anyone?). What you whant is to define your own classes and import the SASS mixins in your own definitions :)

**Comment by michaelBenin** [comment]
Yes, each component has it's own settings. 

It's over 1000 lines long no?

**Comment by gakimball** [comment]
@proimage We haven't come up with an intelligent way to update the settings file. We'd need to do a Git-style intelligent merge of file changes, but *only* override things that are commented out. If a value is uncommented, it means you're using it. Or we could lazily drop all of the new variables at the bottom of the file.

Hopefully better documentation will help make using variables easier. For example, we could be documenting what version numbers introduced which variables. We should also be making it more obvious in our changelogs.

**Comment by doertedev** [comment]
Well since this file is 99% commented out, and one percent kills integration with diff. directory setups (speaking of bower: The loading of the functions file), we could easily:

* drop the documentation
* show the user how to usw mixins
* keep the component-based variables list in their doc site

This is how people got confused in the first place: the assumption that sass-lang is just "hey you can indent, use & for the parent selector and define variables". This is how they just started scribbling around in the _settings.scss

**Comment by hershmire** [comment]
I'd suggest getting rid of vendor prefixes altogether and use [autoprefixer](https://github.com/postcss/autoprefixer) instead.

**Comment by gakimball** [comment]
@hershmire That's the plan. We're also looking into using [css-flip](https://github.com/twitter/css-flip) to make maintaining RTL a little easier.

**Comment by hershmire** [comment]
@gakimball Awesome!

**Comment by adamj88** [comment]
On the subject of modularity, what about splitting each component down into it's own bower package? Inuit does this and it works very well imo.

**Comment by rafibomb** [comment]
@fhunt45 Settings variables had to be added manually anytime a variable was created. We've gotten better at delaying PR's until the settings variables were updated. Now we manually audit this before each release.

In Foundation Apps, we created a way to automatically generate these. It's working well so we'll likely extend it to Sites in the next version.

**Comment by ahmadmilzam** [comment]
Component variants are turned off by default, like buttons. So if I just want buttons, I only get the default button without every different sizes and permutation of them unless i enable them in settings file.

**Comment by gakimball** [comment]
@ahmadmilzam We've talked about this internally, how granular do you see such a feature being? Take buttons, for example, which have a base class, and then modifiers for sizing and coloring. You could make those three things separate imports, but it's more likely that you're using, say, 1/4 colors and 2/4 sizing classes, which means you need a little of both.

**Comment by gakimball** [comment]
Hey guys, we'd like to run some changes by you, related to how component CSS is exported in the Sass.

In Foundation 5, to import a component, you just use `@import`, but if you only want the mixins and not the CSS output, you also dig through the settings file and find an `$include-html-*-classes` variable to set to `false`.

```scss
@import "foundation/components/button";

$include-html-button-classes: false;
```

In Foundation for Apps, we tried a different method, where the handling of exports was done by a single map variable. This eliminates the need to hunt for each component's on/off switch.

```scss
$include-css: (
  global: true,
  button: false,
  // ...
);
```

There's one big issue with both of the above approaches, though, which is that the act of importing (through `@import`) and printing CSS are very tightly coupled. This is why the settings file has to be parsed before the main framework; by the time the compiler reaches `@import "foundation"`, it's already spit out the entire framework. That means nothing about the framework can be changed after the `@import` statement.

In our alpha build of 6.0, we're looking at another approach which would totally separate the acts of importing the framework, and exporting CSS. Basically we just turn every component's CSS output into its own mixin. Your main stylesheet would look like this:

```scss
// No CSS exported
@import "foundation";

// Component CSS exported here
@include foundation-global-styles;
@include foundation-grid-classes;
@include foundation-typography;
```

This maps closer to the user's understanding of Sass, which is that CSS output is handled by statements in your base stylesheet, not by variables in a file that is actually optional. Being a mixin also makes the CSS itself more portable; you can change its position in the cascade, or nest it inside another selector.

This also allows you to change settings variables *after* the framework is imported instead of before, which avoids weird problems the settings file has, like needing to import specific pieces of the framework (e.g. functions) before importing the entire thing.

Lastly, if we wanted to, we could also add parameters to the mixins, allowing you to change the names of every class that's output. This would allow us to make class name customizable, without needing to clutter the settings file with tons of variables, or without the user needing to recreate our class structure with mixins (if all you wanted to do was change the names).

```scss
@include foundation-grid-classes(
  $row: 'hilera',
  $column: 'columna'
);
```

The main downside of this approach over the map approach we use in Foundation for Apps, is that with the map, you get this nice list of key/value pairs showing which components are turned on and off. The full scope of the framework is there for you. With the above approach, on the other hand, CSS is disabled by commenting out or removing lines of code, and if you remove the line of code entirely, to re-add it you'll have to go find in the documentation what it's called.

Let us know what you guys think! And if you have any other concerns with the Sass/CSS, as usual we're all ears :)

**Comment by Martinspire** [comment]
I think you should not only focus on how to include different modules, but also on how people will want to customize (and keep that separate from updating Foundation) so we can use our own colors and styles on top of Foundation. The main issue i have with Foundation and Bootstrap at the moment is that customization is pretty much stuck to a single version and difficult to upgrade automatically.

I need more time to read up some stuff to give you guys some proper feedback but the first suggestion on how to set it up seems all right.

And i don't really understand why you would remove comments in development. Thats for distribution.

**Comment by micahgodbolt** [comment]
I really like this approach, you are in essence turning all of Foundation into a single non-printing Sass partial, with a TON of mixins. Now you can @include a feature rather than @importing one. 

I'm a little curious about performance implications as you will always be loading the entire library into memory, even if you aren't going to be printing any. But past that, mixins are quite powerful, and in essence give you similar functionality in that they print a chunk of code into your CSS

Possible Pros:
- Flat Name Space: No need for @import "this/is/in/_here.scss"
- The ability to pass variables into a component - @incluede foundation-typography('16px', '#646464, $vrhythm: '1.2/1.8')
- Scoped variables within each mixin (override global vars without affecting var in global space)

Cons:
- Performance?
- Globbing - We can't programmatically call mixins (@include #{&name}) so every mixin would have to be called individually, or you'd need mixins that call larger groups of mixins. In a perfect world we'd have a map of components to import and variables to be passed into them.
- Import Once - Compass has an "import-once" feature, but by nature, mixins can be imported as many times as you want. On the other hand, it would be possible to limit multiple imports by keeping a map of already included mixins, and just not print anything if called a 2nd time....or toss a warning or error.

**Comment by gakimball** [comment]
@micahgodbolt The compile time is a valid concern, but it would still be possible to import individual partials as always. We're also trying to push people more towards libsass anyway, which makes the compile time almost negligible.

Regarding the globbing thing, it would be nice if there was a way to store large chunks of CSS in a variable, so you could effectively spring-load them. Maybe that's also a performance issue, but just some way to programmatically summon big chunks of content, like you're describing.

I want to define a component like this:

```scss
@include exports(button) {
  .button { }
}
```

And then print the CSS inside like this:

```scss
// All of it
@include foundation-css;

// Just certain components
@include foundation-css(global type button);
```

This makes it easier to do things programmatically, and with just one line of code.

There's an open Sass issue (sass/sass#1582) related to a feature that would allow for this use-case, but it sounds like it won't make it into version 4.0 of Sass. And we'd need it in libsass anyway :P

We can definitely create shorthands for multiple components, though. Like `@include foundation-all-classes` to just get everything. You could also write a *really long* mixin that parses a list of components like above and includes the right mixins, with like 20 `@if` statements.

Certain components will also have sub-components, and we can create individual mixins along with shorthands. For example, typography is now split into four Sass files: base (only tag selectors), helper classes, alignment classes, and print styles. You can import each one individually or use one shorthand mixin to get all four.

**Comment by blackfalcon** [comment]
There is something to this technique. If I had the following code in a imported partial: 

```scss
$include-classes: true !default;

@mixin color-set ($arg: orange) {
  color: $arg;
}

@if $include-classes {
  .ball {
    @include color-set(white);
    stitches: red;
    material: leather;
    status: hard;
  }
  .bat {
    material: wood;
    length: 4ft;
    @include color-set(brown);
  }
}

```

Then in the file that is importing this partial, I had the following: 

```scss
$include-classes: false;
@import "imports";

.car {
  @include color-set(red);
  wheels: 4;
  doors: 3;
}

$include-classes: true;
@import "imports";
```

The first import would allow for the mixin to pass through and the output CSS would not be in the file, but the second import would reset the flag to true and then the CSS would be imported and output. 

```css
.car {
  color: red;
  wheels: 4;
  doors: 3; }

.ball {
  color: white;
  stitches: red;
  material: leather;
  status: hard; }

.bat {
  material: wood;
  length: 4ft;
  color: brown; }
```

I am not advocating this in a framework scenario as I feel strongly that the CSS output and the Sass tools should be keep apart from each other as not to run into this situation. But, in a case where you have mixed concerns in a single file that is being imported, this is a good solution. That is until the new Sass feature to cherry pick what is imported is completed.

**Comment by zzramesses** [comment]
I love this approach.

**Comment by andy-saint** [comment]
Here's an example of the "lego" approach to building components via basscss.

* [Page Layout](http://www.basscss.com/docs/guides/layout/)
* [UI Elements](http://www.basscss.com/docs/guides/ui/)

Is this along the lines of what you have planned for v4? If so, that's awesome :)

I'd also like to suggest using [postcss](https://github.com/postcss/postcss) in the build process, it makes things like autoprefixing, rem with pixel fallback and RTL support a breeze.

**Comment by gakimball** [comment]
@andy-saint We're definitely trying to be smarter with how components are structured. For example, side nav, sub nav, icon bar, and top bar's dropdowns are being combined into one component called menu bar. Every style of navigation uses the same markup to keep things consistent, and to make it easy to shift around nav styles.

Similarly, panel and alert are being merged into one component, split button is being rolled into button group, and Magellan's navigation and sticky header features are being separated into separate components. So some components are becoming smaller and more reusable, while others are being combined to cut down on extra CSS.

As far as post-processing, we'll be using Autoprefixer for sure. As for RTL, we had looked at rtlcss and css-flip, but we need the flipping functionality at the Sass level, not the CSS level, so we'll probably continue with a method similar to what we do now.

**Comment by florianbouvot** [comment]
+1 for class selectors
+1 for BEM

**Comment by justrhysism** [comment]
Can we please drop the "2 spaces". Use tabs and set your editors to display at 2 spaces. Then those of us who prefer to read at "4 spaces" can still read it as we like.

**Comment by proimage** [comment]
+1 for converting from spaces to tabs for indentation.

**Comment by joelkinzel** [comment]
Can't you just configure your editor? Most of them expand tabs into spaces anyway, its just a matter of how many (2 or 4 most times).

**Comment by justrhysism** [comment]
But that's backwards. It should be the other way around.

**Comment by proimage** [comment]
Using tabs instead of spaces also shrinks files a bit. I know minifying does that and more anyway for production, but still... :stuck_out_tongue:

**Comment by LukePettway** [comment]
"The goal for version 6 is 100%" - This is exciting! I am very stoked to see this large leap in accessibility.

**Comment by zurbrandon** [comment]
Thanks for the feedback, all! We've looked it over and are really rocking away at getting Foundation 6 out very soon!
